examples/event_loop/src/event_loop_example1.cpp

```diff
@@ -20,101 +20,106 @@
+    }
 
-  // Create publishers for each image source
-  for (int i = 0; i < totalCameras; i++) {
-    publishers.push_back(
-        eventLoopProvider.getPublisher<kpsr::vision_ocv::ImageData>(
-            prefixChannel + std::to_string(i), 0,
-            imageCreator.initializerFunction,
-            imageCreator.eventClonerFunction));
-  }
+    // Create image service for each image source
+    std::vector<std::unique_ptr<kpsr::vision_ocv::FileImageStreamingService>> readFromDiskServices;
+    for (int i = 0; i < totalCameras; i++) {
+        readFromDiskServices.emplace_back(
+            new kpsr::vision_ocv::FileImageStreamingService(nullptr,
+                                                            publishers[i],
+                                                            inputFolderName,
+                                                            true,
+                                                            width,
+                                                            height,
+                                                            "camera_link",
+                                                            "streaming_from_rgb_camera"));
+        readFromDiskServices[i]->startup();
+    }
 
-  // Create image service for each image source.
-  for (int i = 0; i < totalCameras; i++) {
-    readFromDiskServices.emplace_back(
-        new kpsr::vision_ocv::FileImageStreamingService(
-            nullptr, publishers[i], inputFolderName, true, width, height,
-            "camera_link", "streaming_from_rgb_camera"));
-    readFromDiskServices[i]->startup();
-  }
+    // Each service is run separately
+    std::vector<std::thread> serviceRunnerThreads;
+    for (int i = 0; i < totalCameras; i++) {
+        serviceRunnerThreads.emplace_back([readTimes, i, &readFromDiskServices]() {
+            for (int j = 0; j < readTimes; j++) {
+                spdlog::info("Service Runner {}. Iteration {}. Thread-id: {}",
+                             i,
+                             j,
+                             getPrintableThreadId());
+                readFromDiskServices[i]->runOnce();
+                std::this_thread::sleep_for(std::chrono::milliseconds(5));
+            }
+        });
+    }
 
-  // Each service is run separately
-  std::vector<std::thread> serviceRunnerThreads;
-  for (int i = 0; i < totalCameras; i++) {
-    serviceRunnerThreads.emplace_back([readTimes, i, &readFromDiskServices]() {
-      for (int j = 0; j < readTimes; j++) {
-        spdlog::info("Service Runner {}. Iteration {}. Thread-id: {}", i, j,
-                     getPrintableThreadId());
-        readFromDiskServices[i]->runOnce();
-        std::this_thread::sleep_for(std::chrono::milliseconds(5));
-      }
-    });
-  }
+    // destruction
+    for (int i = 0; i < totalCameras; i++) {
+        serviceRunnerThreads[i].join();
+        readFromDiskServices[i]->shutdown();
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
Fix premature listeners’ removing:

```text
[2023-09-04 13:39:09.185] [info] Listener 179640912. Thread-id: 140513523316480
```

(id should be 0--2)

Reply:
This was also a problem in the workshop, but I seem to recall we left it as is for simplicity sake.

I guess what you would do would be to put a while loop before the destruction to check that all events were being received.

But I would leave your solution, maybe add a comment saying it's only for code simplicity,